pkg/fqdn/bootstrap/dns_proxy.go

```diff
@@ -0,0 +1,115 @@
+	"github.com/cilium/cilium/pkg/fqdn/messagehandler"
+	"github.com/cilium/cilium/pkg/fqdn/proxy"
+	"github.com/cilium/cilium/pkg/fqdn/re"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+// Cell provides the DNS L7 proxy
+var Cell = cell.Module(
+	"dns-proxy",
+	"Starts the DNS proxy",
+
+	cell.Provide(newFQDNProxyBootstrapper),
+	cell.Provide(newDNSProxy),
+)
+
+type dnsProxyParams struct {
+	cell.In
+
+	Lifecycle         cell.Lifecycle
+	Logger            *slog.Logger
+	DNSRequestHandler messagehandler.DNSMessageHandler
+	EndpointManager   endpointmanager.EndpointManager
+	IPCache           *ipcache.IPCache
+	DefaltProxy       defaultdns.Proxy
+}
+
+// newDNSProxy initializes the DNS l7 proxy.
+func newDNSProxy(params dnsProxyParams) (proxy.DNSProxier, error) {
+	if err := re.InitRegexCompileLRU(params.Logger, option.Config.FQDNRegexCompileLRUSize); err != nil {
+		// can only happen if LRUSize is negative
+		return nil, err
+	}
+
+	// Do not start the proxy in dry mode or if L7 proxy is disabled.
+	// The proxy would not get any traffic in the dry mode anyway, and some of the socket
+	// operations require privileges not available in all unit tests.
+	if option.Config.DryMode || !option.Config.EnableL7Proxy {
+		return nil, nil
+	}
+
+	dnsProxyConfig := dnsproxy.DNSProxyConfig{
+		Logger:                 params.Logger,
+		Address:                "",
+		IPv4:                   option.Config.EnableIPv4,
+		IPv6:                   option.Config.EnableIPv6,
+		EnableDNSCompression:   option.Config.ToFQDNsEnableDNSCompression,
+		MaxRestoreDNSIPs:       option.Config.DNSMaxIPsPerRestoredRule,
+		ConcurrencyLimit:       option.Config.DNSProxyConcurrencyLimit,
+		ConcurrencyGracePeriod: option.Config.DNSProxyConcurrencyProcessingGracePeriod,
+	}
+
+	i := &ipinfo{
+		epm: params.EndpointManager,
+		ipc: params.IPCache,
+	}
+
+	proxy := dnsproxy.NewDNSProxy(dnsProxyConfig, i, params.DNSRequestHandler.NotifyOnDNSMsg)
+	proxy.SetRejectReply(option.Config.FQDNRejectResponse)
+
+	params.DefaltProxy.Set(proxy)
+
+	return proxy, nil
+}
+
+var _ dnsproxy.IPInfo = (*ipinfo)(nil)
+
+type ipinfo struct {
+	epm endpointmanager.EndpointManager
+	ipc *ipcache.IPCache
+}
+
+// lookupEPByIP returns the endpoint that this IP belongs to
+func (i *ipinfo) LookupRegisteredEndpoint(endpointAddr netip.Addr) (endpoint *endpoint.Endpoint, isHost bool, err error) {
+	if e := i.epm.LookupIP(endpointAddr); e != nil {
+		return e, e.IsHost(), nil
+	}
+
+	if node.IsNodeIP(endpointAddr) != "" {
+		if e := i.epm.GetHostEndpoint(); e != nil {
+			return e, true, nil
+		} else {
+			return nil, true, errors.New("host endpoint has not been created yet")
+		}
+	}
+
+	return nil, false, fmt.Errorf("cannot find endpoint with IP %s", endpointAddr)
+}
+
+func (i *ipinfo) LookupSecIDByIP(ip netip.Addr) (secID ipcache.Identity, exists bool) {
```

Comment:
Could just embed the ipc so you don't have to repeat this signature here?

Reply:
good point. I passed an IPC interface directly to the proxy.